File: target_position/src/target_position.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "mobile_manipulator_interfaces/msg/pose.hpp"
#include "mobile_manipulator_interfaces/srv/set_target_position.hpp"

using namespace std::chrono_literals;

/* This example creates a subclass of Node and uses std::bind() to register a
* member function as a callback from the timer. */

class target_server : public rclcpp::Node
{
  public:
    target_server()
    : Node("target_position")
    {
      publisher_ = this->create_publisher<mobile_manipulator_interfaces::msg::Pose>("target_coordinates", 10);
      // service_ = this->create_service<mobile_manipulator_interfaces::srv::set_target_position>("set_target_position",&)
    }

  private:
    // void target_position_callback(this, request: mobile_manipulator_interfaces::srv::SetTargetPosition::Request, responds){

    // }
    // { // defining end-effector target position
    //   auto msg = mobile_manipulator_interfaces::msg::Pose();
    //   msg.position.x = 10.0;
    //   msg.position.y = 20.0;
    //   msg.position.z = 30.0;
    //   msg.orientation.x = 0;
    //   msg.orientation.y = 0;
    //   msg.orientation.z = 0;
    //   msg.orientation.w = 1;

    //   publisher_->publish(msg);
    // }

    rclcpp::Publisher<mobile_manipulator_interfaces::msg::Pose>::SharedPtr publisher_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<target_server>());
  rclcpp::shutdown();
  return 0;
}

File: target_position_py/target_position_py/target_position.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from mobile_manipulator_interfaces.msg import Pose
from mobile_manipulator_interfaces.srv import SetTargetPosition


class TargetPositionNode(Node):

    def __init__(self):
        super().__init__("target_position")
        self.get_logger().info("Target_position_node created.")
        self.pub_ = self.create_publisher(Pose, "target_postion", 10)
        self.timer_ = self.create_timer(0.1, self.publish_target_position)
        self.create_service(SetTargetPosition, "request_target_position", self.target_position_callback)    

    def publish_target_position(self):
        self.msg = Pose()
        self.msg.position.x = 10.0
        self.msg.position.y = 20.0
        self.msg.position.z = 30.0
        self.msg.orientation.x = 0.0
        self.msg.orientation.y = 0.0
        self.msg.orientation.z = 0.0
        self.msg.orientation.w = 1.0
        self.pub_.publish(self.msg)

    def target_position_callback(self, request:SetTargetPosition.Request, response: SetTargetPosition.Response):
        self.get_logger().info("request for target position")
        response.target_position = self.msg
        return response


def main(args=None):
    rclpy.init(args=args)
    node = TargetPositionNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()